src/Dialect/ONNX/AdditionalONNXOps.td

```diff
@@ -140,16 +140,51 @@ def ONNXCustomOp:ONNX_Op<"Custom",
      DeclareOpInterfaceMethods<ShapeHelperOpInterface>]> {
   let summary = "ONNX Custom operation";
   let description = [{
+    This operation is not part of the standard and was added to support call
+    to any user defined external function.
     Allow call-out to a user defined operation. A single attribute
-    is a string which names the operation, other inputs are
-    passed to the user operation.
+    is a string which names the operation, other inputs are passed to the user
+    operation.
     The number of inputs and outputs can vary.
-
-    This operation is not part of the standard and was added to assist onnx-mlir.
+    Since the semantics of this op is not defined to compiler, some attributes
+    are introduced to guide shape inference:
+      'output_type':
+           Optional. The element type for the output tensor. Optional. If not
+           specified, follow the shape infer.
+      'shape_infer_pattern':
+           Optional. Specify how to propagate the shape info from inputs to
```

Comment:
This param is reported as optional here. But in shape inference, there seems to be an assert (line 773). 
We are now also supporting a new approach where shape inference is always fine (aways return success) because we may do shape inference and then change an op to something else. It's only when the shape inference result is actually used that we test if it is implemented. Unimplemented shape inference are tested using the 
```
  virtual bool isImplemented() { return true; }
```
which is default true, but should obviously be redefined here to satisfy the condition that `shape_infer_pattern` is defined, since it appears to be necessary to lowering.

Reply:
I'd like customOp to pass shape inference if shape inference pattern is not provided, and only get assert error when tensor with dynamic shape is to be allocated at lowering. This check should be in lowering Onnx.Call to krill.call. 
Back to shape inference, customOp may behave as three types, ONNXUnaryOpShapeHelper, ONNXBroadcastOpShapeHelper and ShapeHelper that simply pass (I do not see this type of subclass of ShapeHelper yet). In this PR, I will remove that assertion.

Reply:
please address comment above